neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsListIterator.java

```diff
@@ -0,0 +1,591 @@
+     */
+    private boolean hasRemainingValues(IterationHelper helper) {
+        checkState(nonNull(values), "The current feature is neither multi-valued or a feature map");
+
+        // Not a feature map
+        if (!isFeatureMap) {
+            return helper.hasMoreElements(values);
+        }
+
+        // Filter the entries of the feature map
+        while (helper.hasMoreElements(values)) {
+            FeatureMap.Entry entry = FeatureMap.Entry.class.cast(helper.advance(values));
+            if (isIncludedEntry(entry)) {
+                helper.reverse(values);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Advances the iterator on values for the current multi-valued feature, and returns the value associated to the
+     * specified {@code feature}.
+     *
+     * @param helper  the iteration helper that defines the direction to use
+     * @param feature the multi-valued feature associated with the value
+     *
+     * @return the result of the iteration
+     */
+    @Nonnull
+    private IterationResult<E> getValue(IterationHelper helper, EStructuralFeature feature) {
+        checkState(nonNull(values), "The current feature is neither multi-valued or a feature map");
+
+        final Object value = helper.advance(values);
+        final IterationDirection direction = helper.direction();
+
+        return isFeatureMap
+                ? new IterationResult<>(direction, FeatureMap.Entry.class.cast(value))
+                : new IterationResult<>(direction, value, feature);
+    }
+
+    // endregion
+
+    // region Fast preparation
+
+    /**
+     * Advances this iterator at the specified {@code index}.
+     *
+     * @param index the desired position of this iterator
+     *
+     * @return this iterator
+     *
+     * @throws IllegalStateException  if the iterator is already in use
+     * @throws NoSuchElementException if {@code index > size}
+     * @see List#listIterator(int)
+     */
+    @Nonnull
+    public ContentsListIterator<E> fastAdvance(int index) {
+        checkState(cursor == 0, "Iterator already in use");
+
+        while (cursor < index && features.hasNext()) {
+            EStructuralFeature feature = features.next();
+            currentFeature = feature;
+            cursor += fastPrepare(feature, index);
+        }
+
+        if (index != cursor) {
+            throw new NoSuchElementException();
+        }
+
+        return this;
+    }
+
+    /**
+     * Analyzes the {@code feature} in order to prepare the value at the {@code expectedIndex}.
+     * The returned value can be either the size of the feature or the index of the element if the {@code expectedIndex}
+     * has been reached.
+     *
+     * @param feature       the feature
+     * @param expectedIndex the index to reach
+     *
+     * @return the number of steps taken during this preparation
+     */
+    @Nonnegative
+    // TODO Checks the behavior with feature maps
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)